### ISSUE TITLE
Add option to server startup to take a checkpoint using CRIU library

### ISSUE THREAD

**Issue by tjwatson** [OpenLiberty/open-liberty#16578]
A new option is needed to launch a liberty server and do a checkpoint of the process.

The initial design is to add a --checkpoint option that takes a value indicating the point in server startup to take the checkpoint.  The following is an initial list of possible values.  This is subject to change as we investigate the usefulness and feasibility of implementing each value.

- `kernel` - perform a checkpoint just before the feature manager processes the configured features.
- `feature` - perform a checkpoint just after all the feature bundles have been activated. The application manager will not have processed anything for the installed applications yet
- `app` - perform a checkpoint after all configured applications are ready to serve (or a timeout has occurred).
- `server` - perform a checkpoint after the server says it is started.

It is unclear if the difference between `app` and `server` is worth it.  The motivation for `app` vs. `server` is that the http ports will not be open before `server`.

Once the checkpoint is done the process will terminate immediately.  This termination will be done by `criu` while it has the process frozen for the purposes of doing the checkpoint.  The advantage of using `criu` in this way is that it will prevent any chance of Liberty changing the resources associated with the checkpoint post checkpoint.

In order to allow a restore operation to happen multiple times against the same server configuration it will be necessary to backup the workarea folder for the server.  Without doing this it is possible the restored process will modify the workarea in ways that will prevent additional restores to happen from the same state.

In the long run we will want to have a way switch the workarea during restore such that any operations post restore do not modify the original workarea used in the checkpoint.

Once a checkpoint has been performed the checkpoint state of the server process can be restored by using the `restore` option for the `server` command.  This will ensure the backup workarea is restored and then then necessary `criu` restore command will be run the resume the checkpoint server process.

For example:

`server start myServer --checkpoint=feature`
`server restore myServer`

At this point the server process will be restored and the startup process will be resumed.  At this point the server process will behave as normal and can be stopped and restored again:

`server stop myServer`
`server restore myServer`

**Comment by tjwatson** [comment]
This is part of issue #16384

**Comment by tjwatson** [comment]
To start with we will support the --checkpoint option for the following options:

- features - perform a checkpoint just after all the feature bundles have been activated. The application manager will not have processed anything for the installed applications yet
- applications - perform a checkpoint after all configured applications are ready to serve (or a timeout has occurred).  No http ports will be open yet

For example:
```
server start myServer --checkpoint=features
server start myServer --checkpoint=applications
```
We could add a final checkpoint option for the overall `server` that would be after the channel framework has opened the http ports for service.

**Comment by tjwatson** [comment]
I think the initial work for this is done in the `criu` branch.  @sebratton please close if you agree.

**Comment by sebratton** [comment]
This work has been merged into the openLiberty/open-liberty criu branch.
